cli: TestIncludeFiles failed

cli.TestIncludeFiles [failed](https://mesolite.cluster.engflow.com/invocations/default/b3e0a381-4884-46bd-80be-2561537c4b2e?testReportRun=3&testReportShard=8&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on release-26.1 @ [809f97228f29955caac944bd38b875c859b2d634](https://github.com/cockroachdb/cockroach/commits/809f97228f29955caac944bd38b875c859b2d634):

Fatal error:

```
panic: logging already active; first use:
	goroutine 1779828 [running]:
	runtime/debug.Stack()
		GOROOT/src/runtime/debug/stack.go:26 +0x5e
	github.com/cockroachdb/cockroach/pkg/util/debugutil.Stack(...)
		pkg/util/debugutil/debugutil.go:63
	github.com/cockroachdb/cockroach/pkg/util/log.setActive()
		pkg/util/log/clog.go:449 +0x8b
	github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(_, {{{0xc015d01b60, 0x24}, {0x719efd9, 0x1}, {0x719efd9, 0x1}, {0x70c0e05, 0x6}, {0x719efd9, ...}}, ...})
		pkg/util/log/clog.go:275 +0x45
	github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x8006500, 0xc0084c7310}, 0x2, 0x1, 0xf, 0x0?, {0x7323b87, 0x54}, {0xc00bd01db0, 0x1, ...})
		pkg/util/log/channels.go:104 +0x5a5
	github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
		pkg/util/log/channels.go:34
	github.com/cockroachdb/cockroach/pkg/util/log.loggerKvExec.Infof(...)
		bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:6401
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).publishCatchUpBuffer.func1()
		pkg/kv/kvserver/rangefeed/unbuffered_registration.go:305 +0x22d
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).publishCatchUpBuffer(0xc012f26780, {0x8006500, 0xc0084c7310})
		pkg/kv/kvserver/rangefeed/unbuffered_registration.go:332 +0xcb
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).runOutputLoop(0xc012f26780, {0x8006500, 0xc0084c7310}, 0xc0070f8380?)
		pkg/kv/kvserver/rangefeed/unbuffered_registration.go:238 +0x15c
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*ScheduledProcessor).Register.func1.1({0x8007f08?, 0xc00a250640?})
		pkg/kv/kvserver/rangefeed/scheduled_processor.go:377 +0x37
	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x8007f08, 0xc00a250640})
		pkg/util/stop/stopper.go:440 +0x8a
	created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1589481
		pkg/util/stop/stopper.go:438 +0x125
	 [recovered, repanicked]

```
Stack: 

```
goroutine 1579095 [running]:
testing.tRunner.func1.2({0x6b23520, 0xc004393da0})
	GOROOT/src/testing/testing.go:1872 +0x237
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1875 +0x35b
panic({0x6b23520?, 0xc004393da0?})
	GOROOT/src/runtime/panic.go:783 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x3a9
panic({0x6b23520?, 0xc004393da0?})
	GOROOT/src/runtime/panic.go:783 +0x132
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x80061d0, 0xcba2640}, 0x0, {0x71aacdb?, 0x0?}, {0xc00c821390?, 0x70c91d2?, 0x70c91d2?})
	pkg/util/log/logcrash/crash_reporting.go:434 +0x1e8
github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig({{0xc009fca6c0, 0xc0059e45d8, 0xc0059e4610, 0xc0059e4890, 0xc0059e463d, {0x1, 0xc00ea13d80, 0x0, 0xc0059e45d4, 0xc0059e45d5, ...}}, ...}, ...)
	pkg/util/log/flags.go:114 +0x115
github.com/cockroachdb/cockroach/pkg/util/log.(*TestLogScope).SetupSingleFileLogging(0xc00d6aa000)
	pkg/util/log/test_log_scope.go:359 +0x2fc
github.com/cockroachdb/cockroach/pkg/cli.TestIncludeFiles(0xc002c6d180)
	zip_test.go:1451 +0x392
testing.tRunner(0xc002c6d180, 0x7512fa0)
	GOROOT/src/testing/testing.go:1934 +0xea
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1997 +0x465
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestIncludeFiles
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestIncludeFiles176844296
    test_log_scope.go:82: use -show-logs to present logs inline
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestIncludeFiles3640491257
    crash_reporting.go:434: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestIncludeFiles3640491257
    crash_reporting.go:434: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestIncludeFiles176844296
    crash_reporting.go:434: panic: logging already active; first use:
        goroutine 1779828 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:26 +0x5e
        github.com/cockroachdb/cockroach/pkg/util/debugutil.Stack(...)
        	pkg/util/debugutil/debugutil.go:63
        github.com/cockroachdb/cockroach/pkg/util/log.setActive()
        	pkg/util/log/clog.go:449 +0x8b
        github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(_, {{{0xc015d01b60, 0x24}, {0x719efd9, 0x1}, {0x719efd9, 0x1}, {0x70c0e05, 0x6}, {0x719efd9, ...}}, ...})
        	pkg/util/log/clog.go:275 +0x45
        github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x8006500, 0xc0084c7310}, 0x2, 0x1, 0xf, 0x0?, {0x7323b87, 0x54}, {0xc00bd01db0, 0x1, ...})
        	pkg/util/log/channels.go:104 +0x5a5
        github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
        	pkg/util/log/channels.go:34
        github.com/cockroachdb/cockroach/pkg/util/log.loggerKvExec.Infof(...)
        	bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:6401
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).publishCatchUpBuffer.func1()
        	pkg/kv/kvserver/rangefeed/unbuffered_registration.go:305 +0x22d
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).publishCatchUpBuffer(0xc012f26780, {0x8006500, 0xc0084c7310})
        	pkg/kv/kvserver/rangefeed/unbuffered_registration.go:332 +0xcb
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*unbufferedRegistration).runOutputLoop(0xc012f26780, {0x8006500, 0xc0084c7310}, 0xc0070f8380?)
        	pkg/kv/kvserver/rangefeed/unbuffered_registration.go:238 +0x15c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*ScheduledProcessor).Register.func1.1({0x8007f08?, 0xc00a250640?})
        	pkg/kv/kvserver/rangefeed/scheduled_processor.go:377 +0x37
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x8007f08, 0xc00a250640})
        	pkg/util/stop/stopper.go:440 +0x8a
        created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1589481
        	pkg/util/stop/stopper.go:438 +0x125
--- FAIL: TestIncludeFiles (2.31s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=8</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs @cockroachdb/server @cockroachdb/obs-india-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestIncludeFiles.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58187

--- a/pkg/ccl/testccl/authccl/auth_test.go
+++ b/pkg/ccl/testccl/authccl/auth_test.go
@@ -163,7 +163,6 @@ func authCCLRunTest(t *testing.T, insecure bool) {
 		defer sc.Close(t)
 
 		// Enable logging channels.
-		log.TestingResetActive()
 		cfg := logconfig.DefaultConfig()
 		// Make a sink for just the session log.
 		bt := true
@@ -178,7 +177,7 @@ func authCCLRunTest(t *testing.T, insecure bool) {
 		if err := cfg.Validate(&dir); err != nil {
 			t.Fatal(err)
 		}
-		cleanup, err := log.ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+		cleanup, err := log.ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 		if err != nil {
 			t.Fatal(err)
 		}

--- a/pkg/roachprod/logger/log_redirect.go
+++ b/pkg/roachprod/logger/log_redirect.go
@@ -47,7 +47,9 @@ func InitCRDBLogConfig(logger *Logger) {
 	if err := logConf.Validate(nil); err != nil {
 		panic(fmt.Errorf("internal error: could not validate CRDB log config: %w", err))
 	}
-	if _, err := log.ApplyConfig(logConf, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */); err != nil {
+	// Use ApplyConfigForReconfig to handle both initial configuration and
+	// reconfigurations in tests (via TestingCRDBLogConfig) without racing.
+	if _, err := log.ApplyConfigForReconfig(logConf, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */); err != nil {
 		panic(fmt.Errorf("internal error: could not apply CRDB log config: %w", err))
 	}
 	logRedirectInst.logger = logger
@@ -64,8 +66,9 @@ func TestingCRDBLogConfig(logger *Logger) {
 	if logRedirectInst.cancelIntercept != nil {
 		logRedirectInst.cancelIntercept()
 	}
-	log.TestingResetActive()
 	logRedirectInst = &logRedirect{}
+	// InitCRDBLogConfig uses ApplyConfigForReconfig which atomically resets
+	// the active flag, so no separate TestingResetActive() call is needed.
 	InitCRDBLogConfig(logger)
 }
 

--- a/pkg/sql/admin_audit_log_test.go
+++ b/pkg/sql/admin_audit_log_test.go
@@ -24,7 +24,6 @@ import (
 
 func installSensitiveAccessLogFileSink(sc *log.TestLogScope, t *testing.T) func() {
 	// Enable logging channels.
-	log.TestingResetActive()
 	cfg := logconfig.DefaultConfig()
 	// Make a sink for just the session log.
 	bt := true
@@ -39,7 +38,7 @@ func installSensitiveAccessLogFileSink(sc *log.TestLogScope, t *testing.T) func(
 	if err := cfg.Validate(&dir); err != nil {
 		t.Fatal(err)
 	}
-	cleanup, err := log.ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := log.ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/sql/pgwire/auth_test.go
+++ b/pkg/sql/pgwire/auth_test.go
@@ -184,7 +184,6 @@ func hbaRunTest(t *testing.T, insecure bool) {
 		defer sc.Close(t)
 
 		// Enable logging channels.
-		log.TestingResetActive()
 		cfg := logconfig.DefaultConfig()
 		// Make a sink for just the session log.
 		bt := true
@@ -206,7 +205,7 @@ func hbaRunTest(t *testing.T, insecure bool) {
 		if err := cfg.Validate(&dir); err != nil {
 			t.Fatal(err)
 		}
-		cleanup, err := log.ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+		cleanup, err := log.ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 		if err != nil {
 			t.Fatal(err)
 		}

--- a/pkg/util/log/clog_test.go
+++ b/pkg/util/log/clog_test.go
@@ -404,8 +404,7 @@ func TestGetLogReader(t *testing.T) {
 
 	// Validate and apply the config.
 	require.NoError(t, config.Validate(&sc.logDir))
-	TestingResetActive()
-	cleanupFn, err := ApplyConfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanupFn, err := ApplyConfigForReconfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanupFn()
 
@@ -616,8 +615,7 @@ func TestFd2Capture(t *testing.T) {
 	if err := cfg.Validate(&s.logDir); err != nil {
 		t.Fatal(err)
 	}
-	TestingResetActive()
-	cleanupFn, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanupFn, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/util/log/file_log_gc_test.go
+++ b/pkg/util/log/file_log_gc_test.go
@@ -60,8 +60,7 @@ func TestSecondaryGC(t *testing.T) {
 
 	// Validate and apply the config.
 	require.NoError(t, config.Validate(&s.logDir))
-	TestingResetActive()
-	cleanupFn, err := ApplyConfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanupFn, err := ApplyConfigForReconfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanupFn()
 

--- a/pkg/util/log/flags.go
+++ b/pkg/util/log/flags.go
@@ -109,10 +109,45 @@ func ApplyConfig(
 	fileSinkMetricsForDir map[string]FileSinkMetrics,
 	fatalOnLogStall func() bool,
 ) (logShutdownFn func(), err error) {
-	// Sanity check.
-	if active, firstUse := IsActive(); active {
-		reportOrPanic(context.Background(), nil /* sv */, "logging already active; first use:\n%s", firstUse)
-	}
+	return applyConfigInternal(config, fileSinkMetricsForDir, fatalOnLogStall, false /* allowReconfig */)
+}
+
+// ApplyConfigForReconfig is like ApplyConfig but allows reconfiguration even
+// when logging is already active. This is used by TestLogScope to safely
+// reconfigure logging without racing with background goroutines that may be
+// logging concurrently.
+//
+// The key difference from calling TestingResetActive() followed by ApplyConfig()
+// is that this function atomically resets the active flag and proceeds with
+// configuration, eliminating the race window where a background goroutine could
+// set the active flag between the reset and the config application.
+func ApplyConfigForReconfig(
+	config logconfig.Config,
+	fileSinkMetricsForDir map[string]FileSinkMetrics,
+	fatalOnLogStall func() bool,
+) (logShutdownFn func(), err error) {
+	return applyConfigInternal(config, fileSinkMetricsForDir, fatalOnLogStall, true /* allowReconfig */)
+}
+
+func applyConfigInternal(
+	config logconfig.Config,
+	fileSinkMetricsForDir map[string]FileSinkMetrics,
+	fatalOnLogStall func() bool,
+	allowReconfig bool,
+) (logShutdownFn func(), err error) {
+	// Sanity check: either we allow reconfiguration (and reset the active flag),
+	// or we check that logging is not yet active. This is done atomically under
+	// a single lock acquisition to prevent races with concurrent logging.
+	func() {
+		logging.mu.Lock()
+		defer logging.mu.Unlock()
+		if allowReconfig {
+			logging.mu.active = false
+			logging.mu.firstUseStack = nil
+		} else if logging.mu.active {
+			reportOrPanic(context.Background(), nil /* sv */, "logging already active; first use:\n%s", logging.mu.firstUseStack)
+		}
+	}()
 
 	// Our own cancellable context to stop the secondary loggers below.
 	//

--- a/pkg/util/log/flags_test.go
+++ b/pkg/util/log/flags_test.go
@@ -62,8 +62,7 @@ func TestAppliedConfig(t *testing.T) {
 				t.Fatal(err)
 			}
 
-			TestingResetActive()
-			cleanup, err := ApplyConfig(h.Config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+			cleanup, err := ApplyConfigForReconfig(h.Config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -74,4 +73,53 @@ func TestAppliedConfig(t *testing.T) {
 			actual = strings.ReplaceAll(actual, sc.logDir, "TMPDIR")
 			return actual
 		})
+}
+
+func TestApplyConfigRaceCondition(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// This test demonstrates the race condition that exists when using
+	// TestingResetActive() followed by ApplyConfig(). If any logging
+	// happens between these two calls, ApplyConfig() will panic.
+	t.Run("demonstrate panic with old pattern", func(t *testing.T) {
+		defer func() {
+			if r := recover(); r == nil {
+				t.Fatal("expected panic from ApplyConfig when logging is active, but got none")
+			} else {
+				t.Logf("got expected panic: %v", r)
+			}
+		}()
+
+		sc := ScopeWithoutShowLogs(t)
+		defer sc.Close(t)
+
+		cfg := logconfig.DefaultConfig()
+		if err := cfg.Validate(&sc.logDir); err != nil {
+			t.Fatal(err)
+		}
+		TestingResetActive()
+		setActive()
+
+		// Applying the configuration after setting active should panic.
+		_, _ = ApplyConfig(cfg, nil, nil)
+	})
+
+	// This test demonstrates that ApplyConfigForReconfig() can be called
+	// even when logging is already active, without panicking.
+	t.Run("demonstrate reconfiguring multiple times without panic", func(t *testing.T) {
+		sc := ScopeWithoutShowLogs(t)
+		defer sc.Close(t)
+
+		cfg := logconfig.DefaultConfig()
+		if err := cfg.Validate(&sc.logDir); err != nil {
+			t.Fatal(err)
+		}
+		setActive()
+		// Using ApplyConfigForReconfig should succeed, despite logging being active.
+		cleanup, err := ApplyConfigForReconfig(cfg, nil, nil)
+		if err != nil {
+			t.Fatalf("ApplyConfigForReconfig failed: %v", err)
+		}
+		cleanup()
+	})
 }

--- a/pkg/util/log/fluent_client_test.go
+++ b/pkg/util/log/fluent_client_test.go
@@ -63,8 +63,7 @@ func TestFluentClient(t *testing.T) {
 	require.NoError(t, cfg.Validate(&sc.logDir))
 
 	// Apply the configuration.
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanup()
 
@@ -89,7 +88,7 @@ func TestFluentClient(t *testing.T) {
 	msg, err := json.Marshal(info)
 	require.NoError(t, err)
 
-	const expected = `{"c":1,"f":"util/log/fluent_client_test.go","g":222,"l":72,"message":"hello world","n":1,"r":1,"s":1,"sev":"I","t":"XXX","tag":"logtest.ops","v":"v999.0.0"}`
+	const expected = `{"c":1,"f":"util/log/fluent_client_test.go","g":222,"l":71,"message":"hello world","n":1,"r":1,"s":1,"sev":"I","t":"XXX","tag":"logtest.ops","v":"v999.0.0"}`
 	require.Equal(t, expected, string(msg))
 }
 

--- a/pkg/util/log/formats_test.go
+++ b/pkg/util/log/formats_test.go
@@ -60,8 +60,7 @@ func TestFormatRedaction(t *testing.T) {
 							config.CaptureFd2.Enable = false
 							// Validate and apply the config.
 							require.NoError(t, config.Validate(&sc.logDir))
-							TestingResetActive()
-							cleanupFn, err := ApplyConfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+							cleanupFn, err := ApplyConfigForReconfig(config, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 							require.NoError(t, err)
 							defer cleanupFn()
 

--- a/pkg/util/log/http_sink_test.go
+++ b/pkg/util/log/http_sink_test.go
@@ -106,8 +106,7 @@ func testBase(
 	require.NoError(t, cfg.Validate(&sc.logDir))
 
 	// Apply the configuration.
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanup()
 

--- a/pkg/util/log/logtestutils/log_test_utils.go
+++ b/pkg/util/log/logtestutils/log_test_utils.go
@@ -17,7 +17,6 @@ import (
 // InstallLogFileSink installs a file sink for logging tests.
 func InstallLogFileSink(sc *log.TestLogScope, t *testing.T, channel logpb.Channel) func() {
 	// Enable logging channels.
-	log.TestingResetActive()
 	cfg := logconfig.DefaultConfig()
 	// Make a sink for just the session log.
 	cfg.Sinks.FileGroups = make(map[string]*logconfig.FileSinkConfig)
@@ -36,7 +35,7 @@ func InstallLogFileSink(sc *log.TestLogScope, t *testing.T, channel logpb.Channe
 	if err := cfg.Validate(&dir); err != nil {
 		t.Fatal(err)
 	}
-	cleanup, err := log.ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := log.ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/util/log/otlp_client_test.go
+++ b/pkg/util/log/otlp_client_test.go
@@ -209,8 +209,7 @@ func TestOTLPSink(t *testing.T) {
 		require.NoError(t, cfg.Validate(&sc.logDir))
 
 		// Apply the configuration.
-		TestingResetActive()
-		cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+		cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 		require.NoError(t, err)
 
 		for name, test := range tests {
@@ -283,8 +282,7 @@ func TestOTLPSinkHeaders(t *testing.T) {
 		require.NoError(t, cfg.Validate(&sc.logDir))
 
 		// Apply the configuration.
-		TestingResetActive()
-		cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+		cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 		require.NoError(t, err)
 
 		go func() {

--- a/pkg/util/log/redact_test.go
+++ b/pkg/util/log/redact_test.go
@@ -127,12 +127,11 @@ func TestSafeManaged(t *testing.T) {
 				RedactionPolicyManaged = initRedactionPolicyManaged
 			}()
 
-			TestingResetActive()
 			cfg := logconfig.DefaultConfig()
 			if err := cfg.Validate(&s.logDir); err != nil {
 				t.Fatal(err)
 			}
-			cleanupFn, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+			cleanupFn, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 			if err != nil {
 				t.Fatal(err)
 			}

--- a/pkg/util/log/registry_test.go
+++ b/pkg/util/log/registry_test.go
@@ -76,8 +76,7 @@ func TestIterFileSinks(t *testing.T) {
 	require.NoError(t, cfg.Validate(&sc.logDir))
 
 	// Apply the configuration
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanup()
 
@@ -147,8 +146,7 @@ func TestIterHTTPSinks(t *testing.T) {
 	require.NoError(t, cfg.Validate(&sc.logDir))
 
 	// Apply the configuration
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	require.NoError(t, err)
 	defer cleanup()
 

--- a/pkg/util/log/secondary_log_test.go
+++ b/pkg/util/log/secondary_log_test.go
@@ -42,8 +42,7 @@ func installSessionsFileSink(sc *TestLogScope, t *testing.T) func() {
 	}
 
 	// Apply the configuration.
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/util/log/test_log_scope.go
+++ b/pkg/util/log/test_log_scope.go
@@ -161,9 +161,10 @@ func newLogScope(t tShim, mostlyInline bool) (sc *TestLogScope) {
 		// destination directory.
 		cfg := getTestConfig(&sc.logDir, mostlyInline)
 
-		// Switch to the new configuration.
-		TestingResetActive()
-		sc.cleanupFn, err = ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+		// Switch to the new configuration. We use ApplyConfigForReconfig to
+		// atomically reset the active flag and apply the new configuration,
+		// avoiding a race with background goroutines that may be logging.
+		sc.cleanupFn, err = ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 		if err != nil {
 			return err
 		}
@@ -354,9 +355,10 @@ func (l *TestLogScope) SetupSingleFileLogging() (cleanup func()) {
 		panic(errors.NewAssertionErrorWithWrappedErrf(err, "unexpected error in predefined log config"))
 	}
 
-	// Apply the configuration.
-	TestingResetActive()
-	cleanup, err := ApplyConfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
+	// Apply the configuration. We use ApplyConfigForReconfig to atomically
+	// reset the active flag and apply the new configuration, avoiding a race
+	// with background goroutines that may be logging.
+	cleanup, err := ApplyConfigForReconfig(cfg, nil /* fileSinkMetricsForDir */, nil /* fatalOnLogStall */)
 	if err != nil {
 		panic(errors.NewAssertionErrorWithWrappedErrf(err, "unexpected error in predefined log config"))
 	}

